crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -745,7 +752,31 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.infer_body(body);
         for handler in handlers {
             let ast::ExceptHandler::ExceptHandler(handler) = handler;
-            self.infer_optional_expression(handler.type_.as_deref());
+            if let Some(handled_exceptions) = handler.type_.as_deref() {
+                self.infer_expression(handled_exceptions);
+                let invalid_exceptions: Vec<(&ast::Expr, Type)> = {
+                    self.infer_handled_exception_types(handled_exceptions)
+                        .filter_map(|exception| {
+                            if let ExceptHandlerType::Invalid { node, node_ty } = exception {
+                                Some((node, node_ty))
+                            } else {
+                                None
+                            }
+                        })
+                        .collect()
+                };
+                for (node, ty) in invalid_exceptions.iter().copied() {
+                    self.add_diagnostic(
+                        node.into(),
+                        "invalid-except",
+                        format_args!(
+                            "Can only catch a `BaseException` subclass or a tuple of `BaseException` subclasses, \
+                            not an object of type '{}'",
+                            ty.display(self.db)
+                        )
+                    );
+                }
+            }
```

Comment:
It seems silly to collect the invalid nodes into a `Vec<>` only to immediately iterate over them, but I couldn't make the borrow checker happy otherwise (it was complaining about immutable and mutable borrows in the same scope).

Reply:
Yeah that's simply not possible because `add_diagnostic` takes `&mut self` and `Iterator::next` requires a `&self` because of `self.infer_handled_exception_types`.